### PR TITLE
Add pop'n music controller support to LilyPad.

### PR DESCRIPTION
**Pull request by ekudritski**
Added pop'n music controller support.

*Did not chance version and copyright
*Tested only on windows (vc2015)
*linux/Config.cpp not changed

http://forums.pcsx2.net/Thread-POP-N-Music-Controller-Plugin-for-PS1Controller

### CODE REVIEW

File: plugins/LilyPad/Config.cpp [hunk 1]
```diff
@@ -30,6 +30,14 @@
 // Needed to know if raw input is available.  It requires XP or higher.
 #include "RawInput.h"
 
+
+const wchar_t *padTypes[] = {
+	L"Unplugged",
+	L"Dualshock 2",
+	L"Guitar" ,
+	L"pop'n controller"
```
Comment (ssakash) [inline plugins/LilyPad/Config.cpp]: This should be changed to ``Pop'n Music Controller``

File: plugins/LilyPad/LilyPad.rc [hunk 2]
```diff
@@ -138,6 +136,78 @@ BEGIN
...
+BEGIN
+    CONTROL         "",IDC_LIST,"SysListView32",LVS_REPORT | LVS_SHOWSELALWAYS | LVS_NOSORTHEADER | WS_BORDER | WS_VSCROLL | WS_TABSTOP,7,7,183,285,WS_EX_CLIENTEDGE
+    PUSHBUTTON      "Delete Selected",ID_DELETE,7,296,59,15
+    PUSHBUTTON      "Clear All",ID_CLEAR,71,296,56,15
+    PUSHBUTTON      "Ignore Key",ID_IGNORE,132,296,58,15
+	
+	PUSHBUTTON      "Select", ID_SELECT, 262, 22, 43, 15
+	PUSHBUTTON      "Start", ID_START, 306, 22, 43, 15
+
+    PUSHBUTTON      "White L",ID_TRIANGLE, 196,66,43,15
```
Comment (ssakash) [inline plugins/LilyPad/LilyPad.rc]: lines around 147-150 needs proper alignment. also don't leave a space after the elements at comma for all the Resource script codes.

Example:
``"Start", ID_START, 306, 22, 43, 15`` should be changed to ``"Start",ID_START,306,22,43,15"``

File: plugins/LilyPad/LilyPad.rc [hunk 3]
```diff
@@ -138,6 +136,78 @@ BEGIN
...
+    PUSHBUTTON      "Yellow L",ID_CIRCLE,218,48,43,15
+    PUSHBUTTON      "Green L",ID_R1,240, 66,43,15
+    PUSHBUTTON      "Blue L",ID_CROSS,262, 48,43,15
+    PUSHBUTTON      "Red",ID_L1,284, 66,43,15
+    PUSHBUTTON      "Blue R",ID_SQUARE,306, 48,43,15
+    PUSHBUTTON      "Green R",ID_R2,328, 66,43,15
+    PUSHBUTTON      "Yellow R",ID_DPAD_UP,350, 48,43,15
+    PUSHBUTTON      "White R",ID_L2,372, 66,43,15
+
+
```
Comment (ssakash) [inline plugins/LilyPad/LilyPad.rc]: why are there two new lines here ? please remove them.

File: plugins/LilyPad/resource.h [hunk 4]
```diff
@@ -146,7 +147,7 @@
 // 
 #ifdef APSTUDIO_INVOKED
 #ifndef APSTUDIO_READONLY_SYMBOLS
-#define _APS_NEXT_RESOURCE_VALUE        112
+#define _APS_NEXT_RESOURCE_VALUE        114
```
Comment (ssakash) [inline plugins/LilyPad/resource.h]: why not 113 ? that makes it properly ordererd at a arithmetic progression series.

Comment (ssakash) [inline plugins/LilyPad/resource.h]: why not 113 ? that makes it properly ordererd as a arithmetic progression series.

**Comment by ssakash** [comment]
By the way, according to ``git diff HEAD~10..HEAD --check`` there are five trailing white spaces on the code, please remove them.

> plugins/LilyPad/LilyPad.cpp:1215: trailing whitespace.
plugins/LilyPad/LilyPad.rc:28: trailing whitespace.
plugins/LilyPad/LilyPad.rc:33: trailing whitespace.
plugins/LilyPad/LilyPad.rc:39: trailing whitespace.
plugins/LilyPad/LilyPad.rc:147: trailing whitespace.

**Comment by ssakash** [comment]
By the way, I like the order you have used for the Color button controls of Pop'N Music controller on ``IDD_CONFIG_POPN``. I still haven't tested it out on any game yet though, have you made sure that it works fine on the games you have tested ?

I'll probably do a full review / testing later when I get a bit of more time, it's past bed time here. later ;)

**Comment by ssakash** [comment]
One more thing: please squash all the commits into a single one and add a relevant commit message. something like : ``LilyPad: Add support for Pop'N Music Controller``

**Comment by prafullpcsx2** [comment]
@ssakash : Can you send me a windows build of this plugin? It doesn't seem to compile with my Vs2013.

**Comment by ssakash** [comment]
Appveyor automatically generates the build download links : https://ci.appveyor.com/api/buildjobs/n4cihmv599vqh6iu/artifacts/pcsx2-v1.5.0-dev-229-g9eb9374-master-340-vs2015-Win32-AppVeyor.7z

> It doesn't seem to compile with my Vs2013.

It seems to compile fine on Appveyor's VS2013 test and it did also compile fine on my VS2015 community edition, what errors do you get ?

File: plugins/LilyPad/LilyPad.cpp [hunk 5]
```diff
@@ -1212,6 +1212,12 @@ u8 CALLBACK PADpoll(u8 value) {
 				b1 -= ((sum->sticks[0].horiz > 0) << 5);
 				b1 -= ((sum->sticks[0].vert  > 0) << 6);
 				b1 -= ((sum->sticks[0].horiz < 0) << 7);
+				
+				if (config.padConfigs[query.port][query.slot].type == PopnPad) {
+                    //Left, Right and Down are always pressed.
+                    b1=b1 & 0x1f;
+				}
+
```
Comment (ssakash) [inline plugins/LilyPad/LilyPad.cpp]: braces aren't needed here.

File: plugins/LilyPad/Config.cpp [hunk 6]
```diff
@@ -1833,9 +1844,10 @@ INT_PTR CALLBACK GeneralDialogProc(HWND hWnd, unsigned int msg, WPARAM wParam, L
 				selected = 0;
 				ListView_SetExtendedListViewStyleEx(hWndList, LVS_EX_FULLROWSELECT|LVS_EX_DOUBLEBUFFER, LVS_EX_FULLROWSELECT|LVS_EX_DOUBLEBUFFER);
 				SendMessage(hWndList, LVM_SETEXTENDEDLISTVIEWSTYLE, 0, LVS_EX_FULLROWSELECT);
-				SendMessage(hWndCombo, CB_ADDSTRING, 0, (LPARAM) L"Unplugged");
-				SendMessage(hWndCombo, CB_ADDSTRING, 0, (LPARAM) L"Dualshock 2");
-				SendMessage(hWndCombo, CB_ADDSTRING, 0, (LPARAM) L"Guitar");
+                for (int i = 0; i < numPadTypes; i++) {
+                    SendMessage(hWndCombo, CB_ADDSTRING, 0, (LPARAM) padTypes[i]);
```
Comment (ssakash) [inline plugins/LilyPad/Config.cpp]: braces aren't needed here.

**Comment by ssakash** [comment]
I just tested this on ``Pop'n Music 8`` and each of the controls seems to work fine. :+1:

File: plugins/LilyPad/Config.h [hunk 7]
```diff
@@ -23,9 +23,13 @@ extern u8 ps2e;
...
 	DisabledPad,
 	Dualshock2Pad,
-	GuitarPad
+	GuitarPad,
+	PopnPad
 };
 
+const int numPadTypes= 4;
+extern const wchar_t *padTypes[numPadTypes];
+
```
Comment (ssakash) [inline plugins/LilyPad/Config.h]: shouldn't the subscript of array here be ``numPadTypes-1`` ?

File: plugins/LilyPad/Config.cpp (hunk 1 above)
Comment (ekudritski) [inline plugins/LilyPad/Config.cpp]: I was writen like that on the box.

File: plugins/LilyPad/Config.h (hunk 7 above)
Comment (ekudritski) [inline plugins/LilyPad/Config.h]: there is 4 items in that list.

File: plugins/LilyPad/Config.cpp (hunk 1 above)
Comment (ssakash) [inline plugins/LilyPad/Config.cpp]: Hmm, then at least capitalize it and change it to ``Pop'n Controller`` ?

File: plugins/LilyPad/Config.h (hunk 7 above)
Comment (ssakash) [inline plugins/LilyPad/Config.h]: Hmm, I'm not even sure why I commented that above. maybe I ignored the ``DisabledPad`` ? either way please ignore that comment. :)

**Comment by ekudritski** [comment]
I tested it with pop'n music 13 on PCSX2 1.2.1.r5873 and PCSX2 1.4.0-20160105132032-.
I compiled only LilyPad plugin. As i understand its meant to compile only in 32 bit mode.
And DirectX SDK must be installed.
I failed to compile it with VS2012 but maybe i had broken installation. With VS2015 it compiled fine.

File: plugins/LilyPad/Config.cpp (hunk 1 above)
Comment (ssakash) [inline plugins/LilyPad/Config.cpp]: I really think this should be ``Pop'n Music controller`` as that's how the controller is officially stated on the wiki's.

**Comment by ssakash** [comment]
@refractionpcsx2 

no one's interested on merging this ? :P

**Comment by prafullpcsx2** [comment]
I was wondering if someone can implement turntable controller support in lily pad similar to this. That will probably get dj hero working.

**Comment by refractionpcsx2** [comment]
Is this ready for merging? Is everybody happy with it?

File: plugins/LilyPad/LilyPad.rc [hunk 8]
```diff
@@ -13,13 +13,11 @@
...
 
 /////////////////////////////////////////////////////////////////////////////
-// English (U.S.) resources
+// English (United States) resources
 
 #if !defined(AFX_RESOURCE_DLL) || defined(AFX_TARG_ENU)
-#ifdef _WIN32
 LANGUAGE LANG_ENGLISH, SUBLANG_ENGLISH_US
 #pragma code_page(1252)
-#endif //_WIN32
```
Comment (danilaml) [inline plugins/LilyPad/LilyPad.rc]: Is it supposed to be deleted?

Comment (ekudritski) [inline plugins/LilyPad/LilyPad.rc]: It was removed by VS2015. And as far i know this if is always triggered anyway.

Comment (ssakash) [inline plugins/LilyPad/LilyPad.rc]: I'm not sure why it's removed but this file is only compiled on windows so it shouldn't cause any problem I guess. @turtleli  ?

Comment (danilaml) [inline plugins/LilyPad/LilyPad.rc]: Makes sense. Why it was introduced in the first place then?

Comment (turtleli) [inline plugins/LilyPad/LilyPad.rc]: Probably a previous Visual Studio version introduced it and now VS2015 wants rid of it.

I think it's fine.

**Comment by ekudritski** [comment]
@prafullpcsx2 I think dj hero controller uses usb port and needs other plugin to emulate it. I am interested in making beatmania IIDX controller support, but i dont have game, controller nor info yet.

**Comment by ekudritski** [comment]
Also i have about 4 lines unrelated mainly to remove warnings, should i include those to ?

**Comment by ssakash** [comment]
> Also i have about 4 lines unrelated mainly to remove warnings, should i include those to ?

why not ? but make it into a different commit. don't squash it along with the first commit.

**Comment by ekudritski** [comment]
Now all warnings are gone in my vs 2015.

If it compiles then it would be nice if some one could check that this combo box. And makes sure it works properly.
clipboard01

File: plugins/LilyPad/Config.cpp [hunk 9]
```diff
@@ -1493,12 +1501,11 @@ INT_PTR CALLBACK DialogProc(HWND hWnd, unsigned int msg, WPARAM wParam, LPARAM l
 					Binding *b;
 					Device *dev;
 					if (GetBinding(port, slot, index, dev, b, ffb)) {
-						const static unsigned int axisUIDs[3] = {UID_AXIS_NEG, UID_AXIS_POS, UID_AXIS};
+						const static int axisUIDs[3] = {UID_AXIS_NEG, UID_AXIS_POS, UID_AXIS};
 						int uid = dev->virtualControls[b->controlIndex].uid;
```
Comment (turtleli) [inline plugins/LilyPad/Config.cpp]: Maybe it'd be better to change the UID_AXIS definition from `(1 << 31)` to `(1U << 31)`?

Apart from this line, I think the PR is ok.

**Pull request merged**